### PR TITLE
[ROCm] add scratch memory size in AlgorithmDesc

### PR DESCRIPTION
**Pull request by whchung**
Certain algorithms in MIOpen, the DNN library on ROCm platform, requires
scratch memory be allocated in TensorFlow and passed to MIOpen at runtime.

In MIOpen algorithm finding phase, such information would be returned via
MIOpen API and kept in AlgorithmConfig.

In this pull request, changes to AlgorithmConfig and convolutional operators are
added first. Subsequent PR would cover changes for ROCm StreamExecutor.

### CODE REVIEW

**Comment by whchung** [comment]
@jlebar based on our discussion in #20675, this is the another part of proposed interface changes to enable StreamExecutor on ROCm platform

**Comment by whchung** [comment]
@daniellowell @dagamayank for MIOpen algorithm finding strategy. Please help weigh in.

@jlebar Auto-tuning logic works a bit different between CuDNN and MIOpen.

In CuDNN algorithm finding phase, a vector of available algorithms (via `GetConvolveXXXAlgorithms` in `cuda_dnn.cc`  in CUDA StreamExecutor) is iterated, and passed into CuDNN to find the best algorithm with scratch memory, and the one without scratch memory.

In MIOpen algorithm finding phase, MIOpen `FindConvolveXXXAlgorithm` in ROCm StreamExecutor is invoked *without* the vector of available algorithms. MIOpen would find the best algorithm based on its pre-computed performance database, and return the size of scratch memory needed for the algorithm back to the client application (TensorFlow).

It's therefore such an member variable is added in this pull request. And this member variable won't be used in CuDNN path.

**Comment by jlebar** [comment]
> this member variable won't be used in CuDNN path.

How is a user supposed to avoid reading Algorithm::scratch_bytes_used() in the CUDA case, observing that it is zero, and then having a bug?

> In MIOpen algorithm finding phase, MIOpen FindConvolveXXXAlgorithm API is invoked in ROCm StreamExecutor, without the vector of available algorithms. MIOpen would find the best algorithm based on its pre-computed performance database, and return the ID of the algorithm, plus the size of scratch memory needed for the algorithm back to the client application (TensorFlow).

If at all possible we should use MIOpen the same way we use cudnn, where we ask MIOpen for a list of algorithms, and then TensorFlow / XLA runs all of the algorithms, choosing the fastest one.

We need the ability to run "non-optimal" algorithms for a variety of reasons, including checking for correctness or having limited scratch memory available.

So MIOpen should support an API that works like the cudnn API works today.

**Comment by whchung** [comment]
> How is a user supposed to avoid reading Algorithm::scratch_bytes_used() in the CUDA case, observing that it is zero, and then having a bug?

@jlebar How do you think if we refactor the logic, and put this scratch memory size field into `AlgorithmDesc`, and ensure the following semantic:

For both CUDA and ROCm paths:
- `AlgorithmConfig::algorithm_()` would contain the algorithm ID which uses scratch memory, plus the size of scratch memory.
- `AlgorithmConfig::algorithm_no_scratch()_` would contain the algorithm ID which doesn't use scratch memory, with the size of scratch memory be 0.

This could be implemented without changing MIOpen API. I'll address your other comment in another reply.

**Comment by whchung** [comment]
> If at all possible we should use MIOpen the same way we use cudnn, where we ask MIOpen for a list of algorithms, and then TensorFlow / XLA runs all of the algorithms, choosing the fastest one.
> We need the ability to run "non-optimal" algorithms for a variety of reasons, including checking for correctness or having limited scratch memory available.
> So MIOpen should support an API that works like the cudnn API works today.

To my understanding a new MIOpen API termed as "direct mode" is under development where it's possible for client applications to specify exact algorithm be used. This should help realize your request to allow XLA and TensorFlow to pick the best/not-so-best algorithm. However this feature is not released yet.

I'd like to ask @daniellowell to weigh in on the availability of this "direct mode". He is in charge of MIOpen API design.

**Comment by jlebar** [comment]
> How do you think if we refactor the logic, and put this scratch memory size field into AlgorithmDesc, and ensure the following semantic: [...]

sgtm!

> To my understanding a new MIOpen API termed as "direct mode" is under development where it's possible for client applications to specify exact algorithm be used. This should help realize your request to allow XLA and TensorFlow to pick the best/not-so-best algorithm. However this feature is not released yet.

OK.

Until direct mode is available, I think we can still implement the same API as we have for cudnn, though?  Just have ROCm return the one (autotuned) algorithm in the list of algorithms.

If the presence of algorithm_no_scratch is a problem for ROCm (because you only get one algorithm with however much scratch required -- take it or leave it), we can try a bigger refactoring that gets rid of that...  Once you're returning the scratch memory requirement in the AlgorithmDescription, then we can simplify this whole thing quite a bit just by making the client preallocate the scratch buffer.  No need to have a ScratchAllocator in these calls at all -- it can just accept the preallocated scratch buffer instead.

**Comment by dagamayank** [comment]
@jlebar 

> If at all possible we should use MIOpen the same way we use cudnn, where we ask MIOpen for a list of algorithms, and then TensorFlow / XLA runs all of the algorithms, choosing the fastest one.

MIOpen attempts to auto-tune all the available algorithms for any particular inputs. This is similar to `cudnnFindConvolutionForwardAlgorithmEx`. But perhaps XLA, TensorFlow is not using that cuDNN API. Different algorithms require different scratch-space sizes ranging between zero to several MBs and hence, MIOpen provides another API to let the user know how much scratch is needed.

> We need the ability to run "non-optimal" algorithms for a variety of reasons, including checking for correctness or having limited scratch memory available.
> Until direct mode is available, I think we can still implement the same API as we have for cudnn, though? Just have ROCm return the one (autotuned) algorithm in the list of algorithms.

I think this is still possible today. One can just pass zero as scratch-space size. However, in this case it is not guaranteed a valid algorithm is available for any and all inputs. In some scenarios scratch is _a_ requirement. This [link](https://github.com/ROCmSoftwarePlatform/MIOpen/wiki/MIOpen-Convolutions) adds some more color to my description.

With `direct` mode, MIOpen will more closely track cuDNN's usage of -
`cudnnGetConvolutionForwardAlgorithm`  -- returns the algo for a particular config
` cudnnGetConvolutionForwardWorkspaceSize` -- returns the scratch size for that config
`cudnnConvolutionForward` -- executes convolution

I think both cuDNN and MIOpen are doing the same thing albeit the implementations are slightly different. If you look at the [description](https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnFindConvolutionForwardAlgorithm) of `cudnnFindConvolutionForwardAlgorithm` -
_This function attempts all cuDNN algorithms (including CUDNN_TENSOR_OP_MATH and CUDNN_DEFAULT_MATH versions of algorithms where CUDNN_TENSOR_OP_MATH may be available) for cudnnConvolutionForward(), using memory allocated via **cudaMalloc()**,_

In this case, cuDNN internally manages all the memory including scratch as needed, whereas, MIOpen explicitly requires users to manage the memory.

**Comment by jlebar** [comment]
> Until direct mode is available, I think we can still implement the same API as we have for cudnn, though? Just have ROCm return the one (autotuned) algorithm in the list of algorithms.

...or if that is hard because you'd need a to pass scratch allocator to the generate-a-list-of-algorithms function (that seems like a bad API), then in the generate-the-list-of-algorithms function, can we run autotuning with zero scratch memory and return that algorithm as our one result?  That will potentially return a slow conv, but it will be functionally correct, which will unblock us.

**Comment by whchung** [comment]
@jlebar I've updated this PR. Now I put the scratch memory size field into AlgorithmDesc, and ensure the following semantic:

For both CUDA and ROCm paths:

AlgorithmConfig::algorithm_() would contain the algorithm ID which uses scratch memory, plus the size of scratch memory.
AlgorithmConfig::algorithm_no_scratch()_ would contain the algorithm ID which doesn't use scratch memory, with the size of scratch memory be 0.

Could you kindly review it once again? Thanks.

**Comment by whchung** [comment]
a gentle ping?

**Comment by jlebar** [comment]
I appreciate the ping, especially because I haven't learned how to juggle my reviews in github, so there's a real possibility I'll miss pending reviews.  But to set expectations, sometimes it's going to take more than a day to get back to you.  I really want it XLA:AMDGPU to happen, and I'm doing my best trying to keep up, but...I have other things I also have to do.  :-/

If not getting a review really quickly blocks you, we should try to come up with a workflow that lets you make progress without relying on super-fast reviews from my side.  I'm not sure what changes we'd make, but happy to take ideas.

**Comment by whchung** [comment]
@jlebar Thanks for prompt reply. I'm not blocked yet as I still have quite a few PRs under preparation for GPU common runtime, operator implementations, and XLA.

**Comment by jlebar** [comment]
@timshen91 you've been looking at this code lately, would you mind doing a review of this (including double-checking my design suggestion)?

**Comment by timshen91** [comment]
@jlebar , I can take this PR.

IIUC cuDNN also supports "given scratch space limit, find the algorithm" [1]. Today, cuda_dnn.cc gets the scratch size limit [2] by calling `scratch_allocator->GetMemoryLimitInBytes(stream)`.

For supporting MIOpen, Instead of propagating the scratch limit through AlgorithmDesc, maybe try to set it in the scratch allocator?

From design's perspective, I think we should have two distinct types:
a) a type that describes the constraint users provide, for the DNN library to suggest an algorithm.
b) a type that describes an exact algorithm either picked by the user, or suggested by the DNN library.

Unfortunately AlgorithmDesc is (b) and a little bit of (a) [3]. But I'm leaning towards separating (a) from (b). Adding scratch space limit to AlgorithmDesc makes them more mixed.

[1] https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnConvolutionBwdDataPreference_t
[2] https://github.com/tensorflow/tensorflow/blob/f55028af4861bc78516164975a43f259507adf60/tensorflow/stream_executor/cuda/cuda_dnn.cc#L2127
[3] https://github.com/tensorflow/tensorflow/blob/f55028af4861bc78516164975a43f259507adf60/tensorflow/stream_executor/cuda/cuda_dnn.cc#L2123

**Comment by whchung** [comment]
@timshen91 Thanks for the review. The idea using `ScratchAllocator` is an interesting one and let me explore this. If it works we won't need any change in `AlgorithmDesc`.

**Comment by whchung** [comment]
@timshen91 Please find the revised PR. Instead of changing `AlgorithmDesc` I now add one member variable `best_algorithm_scratch_size` in `AlgorithmConfig`. It would be used to keep the size of scratch memory for the best algorithm with scratch memory. Convolution operators in TF have also been changed to see show how the new interface is used.

Please notice `best_algorithm_scratch_size` is set in this PR but not read actually as it's not required for CuDNN. For MIOpen integration logic the field would be used, and would be submitted in a separate PR after prerequisites such as #20277 is accepted and merged.

**Comment by timshen91** [comment]
Can you clarify that which of the following algorithm_scratch_size_ is?
* The scratch size required by algorithm_. It's passed from the DNN library to the user.
* The memory limit of scratch allocation required by the user. It's passed from the user to the DNN library.

Honestly I'm not sure which one is it in the context...

**Comment by whchung** [comment]
@timshen91 it would be:

The scratch size required by algorithm_. It's passed from the DNN library to the user.

**Comment by timshen91** [comment]
Ah I see, sorry for the confusion. I was thinking about the other one all the time.

From my understanding, AlgorithmConfig is the one users send to the DNN library (e.g. ThenConvolveWithAlgorithm), and AlgorithmDesc is the one DNN sends to the user (e.g. GetConvolveAlgorithms), so AlgorithmDesc is a better place for describing "The scratch size required by an algorithm".

It's also implementable in cuDNN (cudnnGetConvolutionForwardWorkspaceSize), and I think we should implement that too, maybe in a separate commit.

**Comment by whchung** [comment]
@timshen91 Thanks for the comments. It seems you are leaning toward getting the field added in `AlgorithmDesc`. In fact that was what the previous version of the PR looked like. :)  Let me try revive it.

**Comment by timshen91** [comment]
Yes, and sorry about that. :)

After that, do you plan to use cuDNN's cudnnGetConvolutionForwardWorkspaceSize() to populate the scratch size in AlgorithmDesc? So that the scratch size is always valid, not like 0 sometimes depending on the DNN library.

**Comment by jlebar** [comment]
We use AlgorithmDesc when launching convolutions as well.

Do I have to pass an AlgorithmDesc with the correct scratch requirement on
pain of fatal error?  Or is it ignored?  How do we propose this works?

**Comment by timshen91** [comment]
> In algorithm finding stage (conv_ops.cc, conv_grad_*.cc), use scratch_allocator.TotalByteSize() to populate the field inside AlgorithmDesc, which is embedded in ProfileResult to determine the best algorithm.

Does that mean the scratch size field carries two meanings?
* `AlgorithmConfig(...).algorithm_.algorithm_scratch_size_` is set to `scratch_allocator.TotalByteSize()`, meaning users intend to put a limit on the total scratch for use, this is passed from the user to the DNN library.
* `AlgorithmDesc(...).algorithm_scratch_size_` means the actual scratch memory needed by the very algorithm it keeps. This is passed from the DNN library to the users.

If that's the case, I don't like the dual meaning of algorithm_scratch_size_. If we need both kinds of variables, we can create two of them, and handle them consistently.

**Comment by whchung** [comment]
My intention is to keep the new field only be set by the DNN library and never by users. To my understanding scratch_allocator.TotalByteSize() is actually set by the DNN library?

**Comment by timshen91** [comment]
> To my understanding scratch_allocator.TotalByteSize() is actually set by the DNN library

`scratch_allocator.TotalByteSize()` is set by the user, as `scratch_allocator` is created by the user to allocate scratch memory. They can be overridden with arbitrary user logic [1].

[1] https://github.com/tensorflow/tensorflow/blob/4764bf2986e2779d5c80b5aca08d72d5c878818b/tensorflow/stream_executor/scratch_allocator.h#L48

**Comment by timshen91** [comment]
I see, yes `scratch_allocator.TotalByteSize()` records the memory size allocated by cuDNN. I think that's the way to go. For a while I mistakenly thought `TotalByteSize()` as `GetMemoryLimitInBytes()`.

Can you also add a TODO that the ProfileResult scratch memory usage should be populated by DNN library itself, not user code like `conv_ops.cc`, `conv_grad_*.cc`?

**Comment by whchung** [comment]
@timshen91 sorry it took me a while to update this PR. was out of town for a couple of days and I was also more focused on #20277.

Now the PR has been updated so the new field to track the size of scratch memory is inside `AlgorithmDesc`. Also in CUDA StreamExecutor implementation the field is populated inside `AllocateCudnnXXXWorkspace()`. Notice instead of calling `scratch_allocator.TotalByteSize()` I use the result from `cudnnGetXXXWorkspaceSize()` which should attain the same result.

File: tensorflow/stream_executor/cuda/cuda_dnn.cc
Comment (timshen91) [inline tensorflow/stream_executor/cuda/cuda_dnn.cc]: algorithm_desc is an in-out parameter, so it should be a pointer and preferably put in the end (before scratch_allocator).

Same applies to other two functions.

**Review by timshen91 — DISMISSED** [review]
Looks good!
I have one comment. Please change the parameter type and position before checking in.

File: tensorflow/stream_executor/cuda/cuda_dnn.cc
Comment (whchung) [reply tensorflow/stream_executor/cuda/cuda_dnn.cc]: @timshen91 updated the PR according to the comment. please help take a look once again. thanks.

**Review by timshen91 — DISMISSED** [review]

**Review by jlebar — DISMISSED** [review]

**Comment by whchung** [comment]
@timshen91 I addressed the build errors on CI, made sure they build / run tests correctly on my end. Could you review once again and re-initate the tests? sorry it takes that long.

**Review by timshen91 — APPROVED** [review]

**Review by timshen91 — APPROVED** [review]

**Review by jlebar — APPROVED** [review]

**Pull request merged**
